### ISSUE TITLE
Type issue in a protocol method if `...` is missing from the body

### ISSUE THREAD

**Issue by kkom** [microsoft/pyright#6487]
**Describe the bug**

Pyright finds an issue in the following code snippet:

```python
from typing import Protocol

class Repro(Protocol):
    def foo(self) -> str:
        """docblock"""
```

```
error: Function with declared return type "str" must return value on all code paths
    "None" is incompatible with "str" (reportGeneralTypeIssues)
```

Things are fine if the code looks like this:

```python
from typing import Protocol

class Repro(Protocol):
    def foo(self) -> str:
        """docblock"""
        ...
```

However after changes to PIE790 in Ruff v0.1.6, Ruff wants to remove the `...` - resulting the Pyright error.

@zanieb from @astral-sh thinks that nothing in [PEP 544](https://peps.python.org/pep-0544/) requires the `...` and that Pyright shouldn't see a problem here. See https://github.com/astral-sh/ruff/issues/8756#issuecomment-1817567945

@erictraut – what's your view? It would be great if Pyright and Ruff were to agree on what's legal here.

**VS Code extension or command-line**

Pyright 1.1.336 CLI

**Comment by erictraut** [comment]
Protocols are allowed to provide default implementations of methods, and classes that derive from these protocols inherit these implementations. A method that does not include a `...` placeholder is assumed by pyright to be a method implementation, and it is type checked accordingly. This has been well established by pyright for several years now. Ruff should be updated according to this convention. I'll also respond to the thread in the ruff issue tracker.

**Comment by kkom** [comment]
Thank you for the quick response Eric! This makes sense, and I hope Ruff resolves it soon.

**Comment by sg495** [comment]
Just making a note that this behaviour is not (yet) aligned with Pylint, which raises `W2301:unnecessary-ellipsis]`.

**Comment by kalfa** [comment]
aligning with pylint is also important. unless it is envisioned that if one uses pyright should not use pylint or other linters.

```
@runtime_checkable
class MyProto(Protocol):
    """An object able to do stuff (protocol)"""

    def f1(self) -> str:
        """annoys pylint because there is an ellipsis where there is already a docstring
        pyright is ok
        """
        ...

    # annoys pylint because there is no docstring, pyright is ok
    def f2(self) -> str:
        ...
        
    def f2(self) -> str:
        """annoys pyright since there is no ellipsis and thereof it expects an implementation.
        it is considered normal with pylint
        """
```


the only way to make them both happy is


```
class MyProto(Protocol):
    """An object able to do stuff (protocol)"""

    def f1(self) -> Any:
       """both happy"""
```

since Any can be none I guess, and pyright is happy.
